035-rebalance-types-scaling-brokers.md

```diff
@@ -0,0 +1,118 @@
+Using the `StatefulSet` for Kafka brokers, the corresponding broker IDs are implied in one direction (added from the higher one up, removed from the higher one down) and the cluster operator could get them for the rebalancing.
+But on the other side, Strimzi is moving to use the [`StrimziPodSet`](https://github.com/strimzi/proposals/pull/44) and towards a ZooKeeper-less cluster support so that the broker IDs are handled differently and the user has the flexibility to scale down by removing specific brokers or scale up by specifying new broker IDs.
+For this reason the proposal is about having the user to specify the added or removed brokers for the rebalancing.
+If needed, it also provides the flexibility to scale up the cluster but running the rebalancing taking into account only a subset of the newly added brokers and leaving others just for new topics.
+The same could happen by running a rebalance for removing brokers to make them "empty" but then not doing the actual scale down.
+
+#### Rebalance on scale up
+
+When scaling up the cluster, the user can go through the following procedure:
+
+1. The user increases the number of replicas through the `spec.kafka.replicas` of the `Kafka` custom resource.
+2. When the scaling up is done, the user creates a `KafkaRebalance` custom resource with the `spec.mode` field as `add-broker` and the list of the new brokers in the `spec.brokers` field.
+3. The `KafkaRebalanceAssemblyOperator` starts the interaction with Cruise Control via the `/add_broker` endpoint for getting an optimization proposal (by using the dryrun feature).
+4. The user accepts the proposal by applying the `strimzi.io/rebalance=approve` annotation on it.
+5. The `KafkaRebalanceAssemblyOperator` starts the interaction with Cruise Control via the `/add_broker` endpoint for running the actual rebalancing.
+
+#### Rebalance on scale down
+
+When scaling down the cluster, the user can go through the following procedure:
+
+1. The user creates a `KafkaRebalance` custom resource with the `spec.mode` field as `remove-broker` and the list of the target brokers to remove in the `spec.brokers` field.
+2. The `KafkaRebalanceAssemblyOperator` starts the interaction with Cruise Control via the `/remove_broker` endpoint for getting an optimization proposal (by using the dryrun feature).
+3. The user accepts the proposal by applying the `strimzi.io/rebalance=approve` annotation on it.
+4. The `KafkaRebalanceAssemblyOperator` starts the interaction with Cruise Control via the `/remove_broker` endpoint for running the actual rebalancing.
+5. When the rebalancing is done, the user can finally scale down the cluster by decreasing the number of replicas through the `spec.kafka.replicas` of the `Kafka` custom resource.
+
+It could happen that before the scaling down but after the rebalancing a new topic is created and replicas are added on the target brokers to remove.
```

Comment:
We are gonna need docs on that. Admins should ensure new topics/partitions are not created during a scaling operations. We should also block other Strimzi CC operations during scale down.